File: dotnet/dojos/dojo1/ThirdTry/ThirdTry/Frame.cs
namespace ThirdTry
{
    public class Frame
    {
        public Frame(int firstBall, int secondBall)
        {
            FirstBall = firstBall;
            SecondBall = secondBall;
        }

        public int FirstBall { get; }
        public int SecondBall { get; }

        public int Score => CalculateScore();
        public Frame Next { get; set; }

        protected virtual int CalculateScore()
        {
            return FirstBall + SecondBall + Bonus();
        }

        private int Bonus()
        {
            var isSpare = FirstBall != 10 && FirstBall + SecondBall == 10;
            var isStrike = FirstBall == 10;
            if (isSpare)
            {
                return NextBallScore();
            }
            if (isStrike)
            {
                return NextBallScore() + SecondNextBallScore();
            }
            return 0;
        }

        private int NextBallScore()
        {
            return Next.FirstBall;
        }

        private int SecondNextBallScore()
        {
            var nextIsStrike = Next.FirstBall == 10;
            var nextIsNotLast = Next.Next != null;
            if (nextIsStrike && nextIsNotLast)
            {
                return Next.Next.FirstBall;
            }
            return Next.SecondBall;
        }
    }
}

File: dotnet/dojos/dojo1/JBG/src/main/java/bowling/fifth/Frame.java
package bowling.fifth;

public class Frame {
    private final int firstBall;
    private final int secondBall;
    private Frame next;

    public Frame(int firstBall, int secondBall) {
        this.firstBall = firstBall;
        this.secondBall = secondBall;
    }

    public int countScore() {
        if (firstBall + secondBall == 10) {
            return firstBall + secondBall + next.firstBall;
        }
        return firstBall + secondBall;
    }

    public void setNext(Frame next) {
        this.next = next;
    }
}
